tablecodec/tablecodec.go

```diff
@@ -236,14 +249,40 @@ func EncodeRow(sc *stmtctx.StatementContext, row []types.Datum, colIDs []int64,
 		values[2*i].SetInt64(id)
 		err := flatten(sc, c, &values[2*i+1])
 		if err != nil {
-			return valBuf, errors.Trace(err)
+			return nil, errors.Trace(err)
 		}
 	}
+	return values, nil
+}
+
+// EncodeRowWithColSizeMap encode row data and column ids into a slice of byte,
+// and record the column value size to colSize map.
+func EncodeRowWithColSizeMap(sc *stmtctx.StatementContext, row []types.Datum, colIDs []int64, valBuf []byte, values []types.Datum, colSize map[int64]int64) ([]byte, map[int64]int64, error) {
+	var err error
+	values, err = getEncodeRowValues(sc, row, colIDs, values)
+	if err != nil {
+		return valBuf, nil, err
+	}
+	valBuf = valBuf[:0]
 	if len(values) == 0 {
 		// We could not set nil value into kv.
-		return append(valBuf, codec.NilFlag), nil
+		return append(valBuf, codec.NilFlag), colSize, nil
 	}
-	return codec.EncodeValue(sc, valBuf, values...)
+	length := len(values) - 1
```

Comment:
Please add some comments here, for example, the row layout. Or the codes here are confusing.

Reply:
Below `for loop` use `i+=2` to grow up, so I use `length := len(values) - 1` to avoid panic when the `len(values)` is odd.  But normally, the `len(values)` should always be even.

Reply:
I see, please comment in the code, thanks.

Reply:
done.